lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,537 @@
+
+    /// Push an element into the batch
+    ///
+    /// This function pushes an element into the batch. Callers must be sure to
+    /// call `has_space` prior to calling this function and receive a positive
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is not sufficient space in the batch
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+    }
+}
+
+#[pin_project]
+pub struct Batcher<St, Prt, T>
+where
+    Prt: Partitioner,
+{
+    /// The timer to maintain periodic flushes from this `Batcher`
+    timer: T,
+    /// The maximum number of items that are allowed per-batch
+    batch_item_limit: usize,
+    /// The total number of bytes a single batch in this struct is allowed to
+    /// hold. If None effectively unlimited.
+    batch_allocation_limit: usize,
+    /// The store of live batches. Note that the key here is an option type,
+    /// on account of the interface of `Prt`.
+    batches: HashMap<Option<Prt::Key>, Batch<Prt::Item>, BuildHasherDefault<XxHash64>>,
+    /// The store of 'closed' batches. When this is not empty it will be
+    /// preferentially flushed prior to consuming any new items from the
+    /// underlying stream.
+    closed_batches: Vec<(Option<Prt::Key>, Vec<Prt::Item>)>,
+    /// The partitioner for this `Batcher`
+    partitioner: Prt,
+    #[pin]
+    /// The stream this `Batcher` wraps
+    stream: St,
+}
+
+impl<St, Prt, T> Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+{
+    pub fn new(
+        stream: St,
+        partitioner: Prt,
+        timer: T,
+        batch_item_limit: NonZeroUsize,
+        batch_allocation_limit: Option<NonZeroUsize>,
+    ) -> Self {
+        Self {
+            batch_item_limit: batch_item_limit.get(),
+            batch_allocation_limit: batch_allocation_limit
+                .map_or(usize::max_value(), NonZeroUsize::get),
+            batches: HashMap::default(),
+            closed_batches: Vec::default(),
+            timer,
+            partitioner,
+            stream,
+        }
+    }
+}
+
+impl<St, Prt, T> Stream for Batcher<St, Prt, T>
+where
+    St: Stream<Item = Prt::Item> + Unpin,
+    Prt: Partitioner + Unpin,
+    Prt::Item: ByteSizeOf,
+    T: Timer,
+{
+    type Item = (Option<Prt::Key>, Vec<Prt::Item>);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            if !this.closed_batches.is_empty() {
+                return Poll::Ready(this.closed_batches.pop());
+            }
+            match this.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    if this.timer.has_elapsed() {
```

Comment:
Just realized: I don't think this will actually work in real async code.  If our inner stream is pending, and we have no closed batches, and the timer hasn't elapsed... nothing is being registered to wake us up.

We probably need `Timer` to have something like `fn poll_elapsed(&mut self, cx: &mut Context) -> Poll<..>` so we can properly forward to the real timer implementations we'll be using behind-the-scenes. For example, this would make it a trivial trait impl to have `tokio::time::Interval` as a timer implementation: https://docs.rs/tokio/1.10.0/tokio/time/struct.Interval.html#method.poll_tick

Reply:
It might also make sense to _only_ have something like `poll_elapsed`.  Right now, this is a bit anti-SRP and not super misuse-resistant: what happens if we forget -- somewhere -- to reset the timer?  Additionally, I believe you can still correctly/accurately apply property testing even with the poll interface.

Reply:
I do believe you are correct.

Reply:
Addressed in 814850ba4